File: packages/ui-compose/ui/src/main/kotlin/com/kinetixui/ui/Progress.kt
package com.kinetixui.ui

import androidx.compose.animation.core.animateFloatAsState
import androidx.compose.foundation.background
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.fillMaxHeight
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.runtime.Composable
import androidx.compose.runtime.getValue
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.res.dimensionResource

/**
 * KinetixProgress — mirrors `packages/ui/src/components/progress.tsx`
 * (Radix `Progress`: `h-2 w-full rounded-full bg-muted` track, `bg-primary`
 * indicator translated by `-(100 - value)%`). Compose's native
 * `fillMaxWidth(fraction)` is the idiomatic, RTL-safe equivalent of that
 * translateX trick, so the indicator is sized directly rather than offset.
 *
 * @param value 0f..100f, same scale as the React `value` prop.
 */
@Composable
fun KinetixProgress(
    value: Float,
    modifier: Modifier = Modifier,
) {
    val colors = KinetixColorScheme.current
    val fraction by animateFloatAsState(
        targetValue = (value / 100f).coerceIn(0f, 1f),
        label = "KinetixProgressFraction",
    )
    val trackHeight = dimensionResource(R.dimen.spacing_2) // 8dp, matches `h-2`
    val shape = RoundedCornerShape(dimensionResource(R.dimen.radius_full))

    Box(
        modifier = modifier
            .fillMaxWidth()
            .height(trackHeight)
            .clip(shape)
            .background(colors.muted),
    ) {
        Box(
            modifier = Modifier
                .fillMaxWidth(fraction)
                .fillMaxHeight()
                .background(colors.primary),
        )
    }
}

File: packages/ui-swiftui/Sources/KinetixUI/Progress.swift
//
// Progress.swift — KinetixProgress.
//
// Mirrors packages/ui/src/components/progress.tsx (Radix `Progress`:
// `h-2 w-full rounded-full bg-muted` track, `bg-primary` indicator). The
// indicator is sized directly as a fraction of the track width rather
// than offset with a translate — the idiomatic, RTL-safe SwiftUI form,
// same call as the Compose port.
//

import SwiftUI

public struct KinetixProgress: View {
    @Environment(\.kinetixColors) private var colors

    /// 0…100, same scale as the React `value` prop.
    private let value: Double

    public init(value: Double) {
        self.value = value
    }

    public var body: some View {
        let fraction = CGFloat(max(0, min(1, value / 100)))
        GeometryReader { geo in
            ZStack(alignment: .leading) {
                Capsule().fill(colors.muted)
                Capsule()
                    .fill(colors.primary)
                    .frame(width: geo.size.width * fraction)
            }
        }
        .frame(height: 8) // h-2
        .animation(.default, value: value)
        .accessibilityValue("\(Int(fraction * 100)) percent")
    }
}
